### PR TITLE
Fix text fragment extractor fallbacks

### PR DESCRIPTION
**Pull request by pokey**
The fallback logic for determining whether to use text-based surrounding pairs or parse-tree-based was broken for text fragment extractors based on next-gen scope handlers (ie using the new `@textFragment` tag in a query file).

This breakage meant that in a string like `"""hello"""` in Python, the surrounding pair finder fell back to text-based, which resulted in it thinking `"hello"` was a string, rather than `"""hello"""`, as would happen if it fell back to parse-tree-based

This PR fixes the issue by unifying the fallback logic between legacy and next-gen text fragment extractors, so that they don't fall out of sync again

See also https://github.com/cursorless-dev/cursorless/issues/1812#issuecomment-1691493746; arguably that will make this PR irrelevant, but until then, it's better to have string not be broken when we migrate a language to use `@textFragment`

Note that the tests in this PR don't actually test the code yet, because Python is still using legacy text fragment extractors. The tests will start biting when this PR is merged into main and then merged into https://github.com/cursorless-dev/cursorless/pull/1862

- This PR is required by https://github.com/cursorless-dev/cursorless/pull/1862

## Checklist

- [ ] I have added [tests](https://www.cursorless.org/docs/contributing/test-case-recorder/)
- [ ] I have updated the [docs](https://github.com/cursorless-dev/cursorless/tree/main/docs) and [cheatsheet](https://github.com/cursorless-dev/cursorless/tree/main/cursorless-talon/src/cheatsheet)
- [ ] I have not broken the cheatsheet

### CODE REVIEW

File: packages/cursorless-engine/src/processTargets/modifiers/surroundingPair/index.ts [hunk 1]
```diff
@@ -140,14 +102,41 @@ function processSurroundingPairCore(
...
+
+    if (textFragmentExtractor == null) {
+      // If the text fragment extractor doesn't exist, or if it explicitly is
+      // set to `null`, then we just use text-based algorithm on entire document
+      return document.range;
+    }
+
+    const selectionWithEditor = {
+      editor,
+      selection: new Selection(range.start, range.end),
```
Comment (josharian) [inline packages/cursorless-engine/src/processTargets/modifiers/surroundingPair/index.ts:131]: i think we have a range.toSelection. although looking at it, it has a boolean arg, ick. (maybe we could make that optional/named and default to false?)

**Review by josharian — APPROVED** [review]
lgtm fwiw

File: packages/cursorless-engine/src/processTargets/modifiers/surroundingPair/index.ts (hunk 1 above)
Comment (AndreasArvidsson) [reply packages/cursorless-engine/src/processTargets/modifiers/surroundingPair/index.ts:131]: I think it's useful to force the user to think about the direction. If you don't care about a direction a selection is not that different from a range.

Comment (josharian) [reply packages/cursorless-engine/src/processTargets/modifiers/surroundingPair/index.ts:131]: sure. but at the call site `.toSelection(false)` is inscrutable.

File: packages/cursorless-vscode-e2e/src/suite/fixtures/recorded/surroundingPair/parseTree/python/changeString.yml [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+  action:
+    name: clearAndSetSelection
+    target:
+      type: primitive
+      modifiers:
+        - type: containingScope
+          scopeType: {type: surroundingPair, delimiter: string}
+  usePrePhraseSnapshot: true
+initialState:
+  documentContents: "\"\"\"hello\"\"\""
```
Comment (AndreasArvidsson) [inline packages/cursorless-vscode-e2e/src/suite/fixtures/recorded/surroundingPair/parseTree/python/changeString.yml:14]: Wow the escaping here is horrid

**Review by AndreasArvidsson — APPROVED** [review]

File: packages/cursorless-engine/src/processTargets/modifiers/surroundingPair/index.ts [hunk 3]
```diff
@@ -73,40 +71,6 @@ function processSurroundingPairCore(
...
-        containingScope[0].contentRange,
-        delimiters,
-        scopeType,
-      );
-      if (surroundingRange != null) {
-        // Found the pair within this text fragment or comment, e.g. "(abc)"
-        return surroundingRange;
-      }
-      // Search in the rest of the file, to find e.g. ("abc")
-      return findSurroundingPairTextBased(
```
Comment (pokey) [inline packages/cursorless-engine/src/processTargets/modifiers/surroundingPair/index.ts:101]: Fwiw the specific bug that this PR is trying to fix was on this line: we were directly falling back to the text-based algorithm when the input range was not in a text fragment, when instead it should have fallen back to parse-tree-based

File: packages/cursorless-vscode-e2e/src/suite/fixtures/recorded/surroundingPair/parseTree/python/changeString.yml (hunk 2 above)
Comment (pokey) [reply packages/cursorless-vscode-e2e/src/suite/fixtures/recorded/surroundingPair/parseTree/python/changeString.yml:14]: haha yeah yaml ftw 😅

Comment (AndreasArvidsson) [reply packages/cursorless-vscode-e2e/src/suite/fixtures/recorded/surroundingPair/parseTree/python/changeString.yml:14]: It's fine either way from me

**Pull request merged**